src/neon/stage1.rs

```diff
@@ -0,0 +1,531 @@
+#[cfg_attr(not(feature = "no-inline"), inline(always))]
+unsafe fn cmp_mask_against_input(input: &SimdInput, m: u8) -> u64 {
+    macro_rules! call {
+        ($imm8:expr) => {
+            vmovq_n_u8($imm8)
+        };
+    };
+    let mask: uint8x16_t = constify_imm8!(m, call);
+
+    let cmp_res_0: uint8x16_t = vceqq_u8(input.v0, mask);
+    let cmp_res_1: uint8x16_t = vceqq_u8(input.v1, mask);
+    let cmp_res_2: uint8x16_t = vceqq_u8(input.v2, mask);
+    let cmp_res_3: uint8x16_t = vceqq_u8(input.v3, mask);
+
+    neon_movemask_bulk(cmp_res_0, cmp_res_1, cmp_res_2, cmp_res_3)
+}
+
+// find all values less than or equal than the content of maxval (using unsigned arithmetic)
+#[cfg_attr(not(feature = "no-inline"), inline(always))]
+unsafe fn unsigned_lteq_against_input(input: &SimdInput, maxval: u8) -> u64 {
+    macro_rules! call {
+        ($imm8:expr) => {
+            vmovq_n_u8($imm8)
+        };
+    };
+    let mask: uint8x16_t = constify_imm8!(maxval, call);
+
+    let cmp_res_0: uint8x16_t = vcleq_u8(input.v0, mask);
+    let cmp_res_1: uint8x16_t = vcleq_u8(input.v1, mask);
+    let cmp_res_2: uint8x16_t = vcleq_u8(input.v2, mask);
+    let cmp_res_3: uint8x16_t = vcleq_u8(input.v3, mask);
+
+    neon_movemask_bulk(cmp_res_0, cmp_res_1, cmp_res_2, cmp_res_3)
+}
+
+// return a bitvector indicating where we have characters that end an odd-length
+// sequence of backslashes (and thus change the behavior of the next character
+// to follow). A even-length sequence of backslashes, and, for that matter, the
+// largest even-length prefix of our odd-length sequence of backslashes, simply
+// modify the behavior of the backslashes themselves.
+// We also update the prev_iter_ends_odd_backslash reference parameter to
+// indicate whether we end an iteration on an odd-length sequence of
+// backslashes, which modifies our subsequent search for odd-length
+// sequences of backslashes in an obvious way.
+#[cfg_attr(not(feature = "no-inline"), inline(always))]
+unsafe fn find_odd_backslash_sequences(input: &SimdInput, prev_iter_ends_odd_backslash: &mut u64) -> u64 {
+    const EVEN_BITS: u64 = 0x5555_5555_5555_5555;
+    const ODD_BITS: u64 = !EVEN_BITS;
+
+    let bs_bits: u64 = cmp_mask_against_input(input, b'\\');
+    let start_edges: u64 = bs_bits & !(bs_bits << 1);
+    // flip lowest if we have an odd-length run at the end of the prior
+    // iteration
+    let even_start_mask: u64 = EVEN_BITS ^ *prev_iter_ends_odd_backslash;
+    let even_starts: u64 = start_edges & even_start_mask;
+    let odd_starts: u64 = start_edges & !even_start_mask;
+    let even_carries: u64 = bs_bits.wrapping_add(even_starts);
+
+    let mut odd_carries: u64 = 0;
+    // must record the carry-out of our odd-carries out of bit 63; this
+    // indicates whether the sense of any edge going to the next iteration
+    // should be flipped
+    let iter_ends_odd_backslash: bool = add_overflow(bs_bits, odd_starts, &mut odd_carries);
```

Comment:
Could we just pull in the rust own function here, it might save us another architecture specific function the same way as `trailing_zeros()` does.